### PR TITLE
Log exceptions as nested array instead of encoded json

### PR DESCRIPTION
**Pull request by icewind1991**
Revived version of #7826

Logreader support: https://github.com/nextcloud/logreader/commit/0f4be889b9603597b7254c7f202a2ce449a7c9fa

I would also like to extend the amount of data we log in a follow up pr (mainly "preview" exception)

### CODE REVIEW

**Comment by MorrisJobke** [comment]
There was one problem ... I only stated that it is because of the type hinting. I guess what we hand in into the log implementations. But this looks good so far in your version. 🤔

**Comment by icewind1991** [comment]
From what I understood the problem was the old `logException` calling `log` which is now type-hinted to make calling it with the nested array invalid. Thus I refactored the logging so `logException` does not need to call `log`

**Review by rullzer — APPROVED** [review]
Fine by me

File: lib/private/Log.php [hunk 1]
```diff
@@ -332,26 +334,35 @@ public function log(int $level, string $message, array $context = []) {
...
+		$app = $context['app'] ?: 'no app in context';
+		$level = $context['level'] ?: Util::ERROR;
+
 		$data = [
+			'CustomMessage' => isset($context['message']) ? $context['message'] : '--',
 			'Exception' => get_class($exception),
 			'Message' => $exception->getMessage(),
 			'Code' => $exception->getCode(),
-			'Trace' => $exception->getTraceAsString(),
+			'Trace' => $exception->getTrace(),
```
Comment (MorrisJobke) [inline lib/private/Log.php]: If this is used passwords are still logged 🙈 

```
{
    "reqId": "xaT0UndUeGLdPMu2jnI6",
    "level": 3,
    "time": "2018-04-09T21:18:30+00:00",
    "remoteAddr": "172.20.0.1",
    "user": "--",
    "app": "index",
    "method": "POST",
    "url": "/index.php/login",
    "message": {
        "CustomMessage": "--",
        "Exception": "InvalidArgumentException",
        "Message": "Username must be string",
        "Code": 0,
        "Trace": [
            {
                "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 166,
                "function": "tryLogin",
                "class": "OC\\Core\\Controller\\LoginController",
                "type": "->",
                "args": [
                    "admin",
                    "password",
                    null,
                    false,
                    "Europe/Berlin",
                    "2"
                ]
            },
            {
                "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 99,
                "function": "executeController",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [
                    {},
                    "tryLogin"
                ]
            },
            {
                "file": "/var/www/html/lib/private/AppFramework/App.php",
                "line": 118,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [
                    {},
                    "tryLogin"
                ]
            },
            {
                "file": "/var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php",
                "line": 47,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
                "args": [
                    "OC\\Core\\Controller\\LoginController",
                    "tryLogin",
                    {},
                    {
                        "_route": "core.login.tryLogin"
                    }
                ]
            },
            {
                "function": "__invoke",
                "class": "OC\\AppFramework\\Routing\\RouteActionHandler",
                "type": "->",
                "args": [
                    {
                        "_route": "core.login.tryLogin"
                    }
                ]
            },
            {
                "file": "/var/www/html/lib/private/Route/Router.php",
                "line": 297,
                "function": "call_user_func",
                "args": [
                    {},
                    {
                        "_route": "core.login.tryLogin"
                    }
                ]
            },
            {
                "file": "/var/www/html/lib/base.php",
                "line": 981,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
                "args": [
                    "/login"
                ]
            },
            {
                "file": "/var/www/html/index.php",
                "line": 37,
                "function": "handleRequest",
                "class": "OC",
                "type": "::",
                "args": []
            }
        ],
        "File": "/var/www/html/core/Controller/LoginController.php",
        "Line": 245
    },
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_3) AppleWebKit/604.5.6 (KHTML, like Gecko) Version/11.0.3 Safari/604.5.6",
    "version": "14.0.0.1"
}
```

**Comment by icewind1991** [comment]
@MorrisJobke fixed

**Comment by MorrisJobke** [comment]
Now it removes all parameters:

```
{
    "reqId": "WW1k9pgFAakfRJvaGxGD",
    "level": 3,
    "time": "2018-04-10T09:12:17+00:00",
    "remoteAddr": "172.20.0.1",
    "user": "--",
    "app": "index",
    "method": "POST",
    "url": "/index.php/login",
    "message": {
        "CustomMessage": "--",
        "Exception": "InvalidArgumentException",
        "Message": "Username must be string",
        "Code": 0,
        "Trace": [
            {
                "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 166,
                "function": "tryLogin",
                "class": "OC\\Core\\Controller\\LoginController",
                "type": "->",
                "args": [
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
                "line": 99,
                "function": "executeController",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [
                    "*** sensitive parameter replaced ***",
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "file": "/var/www/html/lib/private/AppFramework/App.php",
                "line": 118,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->",
                "args": [
                    "*** sensitive parameter replaced ***",
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "file": "/var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php",
                "line": 47,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::",
                "args": [
                    "*** sensitive parameter replaced ***",
                    "*** sensitive parameter replaced ***",
                    "*** sensitive parameter replaced ***",
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "function": "__invoke",
                "class": "OC\\AppFramework\\Routing\\RouteActionHandler",
                "type": "->",
                "args": [
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "file": "/var/www/html/lib/private/Route/Router.php",
                "line": 297,
                "function": "call_user_func",
                "args": [
                    "*** sensitive parameter replaced ***",
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "file": "/var/www/html/lib/base.php",
                "line": 982,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->",
                "args": [
                    "*** sensitive parameter replaced ***"
                ]
            },
            {
                "file": "/var/www/html/index.php",
                "line": 37,
                "function": "handleRequest",
                "class": "OC",
                "type": "::",
                "args": []
            }
        ],
        "File": "/var/www/html/core/Controller/LoginController.php",
        "Line": 243
    },
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.1 Safari/605.1.15",
    "version": "14.0.0.2"
}
```

**Comment by icewind1991** [comment]
missed a strict option :see_no_evil:

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Unit tests fail 🙈

**Comment by icewind1991** [comment]
ci is happy

**Review by MorrisJobke — APPROVED** [review]

**Pull request merged**